third_party/nvfuser/csrc/expr_simplifier.cpp

```diff
@@ -1188,12 +1185,29 @@ bool isMultipleOf(Val* x, Val* y) {
   return sym_algebra::divideFactorized(lhs, rhs) != nullptr;
 }
 
-bool hasCompatibleSign(Val* x, Val* y, const VarInfoMap& var_info) {
-  return isNonNegative(x, var_info) && isNonNegative(y, var_info);
+bool hasCompatibleSign(Val* x, Val* y, const Context& context) {
+  return isNonNegative(x, context) && isNonNegative(y, context);
 }
 
 } // namespace prove
 
+namespace {
+
+// If we can not prove `value` to be zero, then treat it as non-zero, unless we
+// want to preserve error. This is useful for simplifications like (a * b) / b.
+// If we want to preserve error, then we can simplify the above expr to `a` iff
+// we can prove that b is not zero. If we don't care about error case, then
+// having `b` in the denominator already indicates that `b` is non-zero. So we
+// can just do the simplification without worrying about `b` being zero.
+bool notSurelyZero(Val* value, const Context& context) {
```

Comment:
nit: this name doesn't sound very intuitive to me. What this tells is `value` is not proven to be zero, right? It seems it just says it may be or may not be zero. Am I right?

Reply:
Your understanding is correct. The name and comment of this function is very confusing, I renamed it and rewite its comment. See https://github.com/csarofeen/pytorch/pull/2534#discussion_r1122340674